File: src/2025-02-23-deepclaude-try/go/exercises/practice/connect/connect.go
package connect

import "strings"

func ResultOf(lines []string) (string, error) {
	// Handle empty input
	if len(lines) == 0 {
		return "", nil
	}

	// Create grid from input lines
	grid := make([][]rune, len(lines))
	for i, line := range lines {
		trimmed := strings.TrimSpace(line)
		grid[i] = []rune(trimmed)
	}

	// Check O's win (top to bottom)
	hasO := hasWinningPath(grid, 'O', isTopRow, isBottomRow)

	// Check X's win (left to right)
	hasX := hasWinningPath(grid, 'X', isLeftCol, isRightCol)

	// Return winner (O takes precedence if both somehow win)
	if hasO {
		return "O", nil
	}
	if hasX {
		return "X", nil
	}
	return "", nil
}

// Helper functions to check if a position is on an edge
func isTopRow(row, _ int) bool    { return row == 0 }
func isBottomRow(row, cols int) bool { return row == cols-1 }
func isLeftCol(_, col int) bool   { return col == 0 }
func isRightCol(_, col int) bool  { return col == col-1 }

// Checks if player has a winning path from their start edge to their target edge
func hasWinningPath(grid [][]rune, player rune, isStart, isTarget func(int, int) bool) bool {
	if len(grid) == 0 {
		return false
	}
	
	rows, cols := len(grid), len(grid[0])
	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}

	// Six possible neighbor directions in hex grid
	dirs := [][2]int{
		{-1, 0},  // up
		{1, 0},   // down 
		{0, -1},  // left
		{0, 1},   // right
		{-1, 1},  // up-right
		{1, -1},  // down-left
	}

	// Queue for BFS
	var queue [][2]int

	// Add all starting positions to queue
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if isStart(r, c) && grid[r][c] == player {
				queue = append(queue, [2]int{r, c})
				visited[r][c] = true
			}
		}
	}

	// BFS to find path
	for len(queue) > 0 {
		pos := queue[0]
		queue = queue[1:]
		r, c := pos[0], pos[1]

		// Check if we reached target edge
		if isTarget(r, c) {
			return true
		}

		// Check all neighbors
		for _, dir := range dirs {
			nr, nc := r+dir[0], c+dir[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols &&
				!visited[nr][nc] && grid[nr][nc] == player {
				visited[nr][nc] = true
				queue = append(queue, [2]int{nr, nc})
			}
		}
	}

	return false
}

File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/connect/connect.js
//
// This is only a SKELETON file for the 'Connect' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Board {
  constructor(input) {
    if (typeof input === 'string') {
      this.grid = input.split('\n').map(line => line.trim().split(/\s+/));
    } else {
      this.grid = (input || []).map(line => line.trim().split(/\s+/));
    }
  }

  winner() {
    // Check O's win condition (top to bottom)
    const oStarts = [];
    if (this.grid.length === 0) return null;
    const firstRow = this.grid[0];
    for (let col = 0; col < firstRow.length; col++) {
      if (firstRow[col] === 'O') {
        oStarts.push([0, col]);
      }
    }
    if (this.checkWin(oStarts, (row) => row === this.grid.length - 1, 'O')) {
      return 'O';
    }

    // Check X's win condition (left to right)
    const xStarts = [];
    for (let row = 0; row < this.grid.length; row++) {
      if (this.grid[row].length > 0 && this.grid[row][0] === 'X') {
        xStarts.push([row, 0]);
      }
    }
    if (this.checkWin(xStarts, (_, col) => col === this.grid[0].length - 1, 'X')) {
      return 'X';
    }

    return null;
  }

  checkWin(starts, targetCheck, player) {
    const visited = new Set();
    const queue = [...starts];

    for (const [row, col] of queue) {
      const key = `${row},${col}`;
      if (visited.has(key)) continue;
      visited.add(key);
      if (targetCheck(row, col)) return true;

      const neighbors = this.getNeighbors(row, col);
      for (const [nr, nc] of neighbors) {
        if (this.grid[nr][nc] === player && !visited.has(`${nr},${nc}`)) {
          queue.push([nr, nc]);
        }
      }
    }
    return false;
  }

  getNeighbors(row, col) {
    const neighbors = [];
    const dirs = row % 2 === 0 
      ? [ [-1, 0], [-1, 1], [0, -1], [0, 1], [1, 0], [1, 1] ] 
      : [ [-1, -1], [-1, 0], [0, -1], [0, 1], [1, -1], [1, 0] ];

    for (const [dr, dc] of dirs) {
      const newRow = row + dr;
      const newCol = col + dc;
      if (newRow >= 0 && newRow < this.grid.length && newCol >= 0 && newCol < this.grid[newRow].length) {
        neighbors.push([newRow, newCol]);
      }
    }
    return neighbors;
  }
}
